src/clients/blocks.ts

```diff
@@ -1,27 +1,48 @@
 import { gql, GraphQLClient } from 'graphql-request';
 import _ from 'lodash';
 
+import { DBClient, Table } from '../db/db';
+import { Block } from '../types/block';
 import { ChainId } from '../types/chain';
 
 import { EVMClient } from './evm';
 
 const blocksAPI = new GraphQLClient(process.env.BLOCKS_SUBGRAPH_ENDPOINT!);
 
 export type BlocksClient = {
-  fetchBlockTimestamps: (chainId: ChainId, blockNumbers: string[]) => Promise<{ [blockNumber: string]: string }>;
+  fetchBlockTimestamps: (chainId: ChainId, blockNumbers: string[]) => Promise<{ 
+    timestampsByBlockNumber: { [blockNumber: string]: string },
+    newBlocks: Block[] 
+  }>;
 }
 
-const init = async (evmClients: { [chainId in ChainId]: EVMClient }) => {
+const init = async (
+  evmClients: { [chainId in ChainId]: EVMClient },
+  dbClient: DBClient
+) => {
   return {
     fetchBlockTimestamps: async (
       chainId: ChainId,
       blockNumbers: string[],
-    ): Promise<{ [blockNumber: string]: string }> => {
-      let result = {};
+    ): Promise<{ 
+      timestampsByBlockNumber: { [blockNumber: string]: string },
+      newBlocks: Block[] 
+    }> => {
+      let timestampsByBlockNumber: { [blockNumber: string]: string } = {};
+
+      const existingBlocks = await dbClient.getRecords<Block>(Table.blocks, [
+        ['where', ['chainId', chainId]],
+        ['whereIn', ['blockNumber', blockNumbers]]
+      ])
+      console.log(`using ${existingBlocks.length} blocks from cache`)
+      const existingBlocksByBlockNumber = _.keyBy(existingBlocks, 'blockNumber');
+
+      const missingBlockNumbers = blockNumbers.filter(blockNumber => !existingBlocksByBlockNumber[blockNumber]);
+      let newBlocks: Block[] = [];
 
       // only using the graph for Ethereum. All other EVM chains will use RPC calls
       if (chainId === ChainId.ethereum){
-        const blockNumbersForQuery = JSON.stringify(blockNumbers);
+        const blockNumbersForQuery = JSON.stringify(missingBlockNumbers);
```

Comment:
should we deprecate the graph dependency entirely with this PR, even for Ethereum?

Reply:
yeah I think we should! It would be nice to remove that dependency, and have a single code path for all chains. This change should reduce any performance impacts related to doing a query per block